### PR TITLE
Fix recalculation of skip indexes in ALTER UPDATE queries when table has adaptive granularity

### LINKED ISSUE #52652: Code: 33. DB::Exception: Cannot read all marks from file
```sql
2023.07.27 10:54:01.785213 [ 1720 ] {9e777690-69a0-4208-aafc-675961364fbf} <Error> executeQuery: Code: 33. DB::Exception: Cannot read all marks from file skp_idx_regionId_index.cmrk3, marks expected 2 (bytes size 48), marks read 3 (bytes size 72). (CANNOT_READ_ALL_DATA) (version 23.5.4.10) (from 10.244.125.151:44596) (in query: SELECT count() FROM ...), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000e0c3d95 in /usr/bin/clickhouse
1. ? @ 0x00000000146daa67 in /usr/bin/clickhouse
2. DB::MergeTreeMarksLoader::loadMarksImpl() @ 0x00000000146d9f22 in /usr/bin/clickhouse
3. DB::MergeTreeMarksLoader::loadMarks() @ 0x00000000146d80ea in /usr/bin/clickhouse
4. DB::MergeTreeMarksLoader::getMark(unsigned long, unsigned long) @ 0x00000000146d7dd8 in /usr/bin/clickhouse
5. DB::MergeTreeReaderStream::init() @ 0x00000000146ff601 in /usr/bin/clickhouse
6. DB::MergeTreeIndexReader::MergeTreeIndexReader(std::shared_ptr<DB::IMergeTreeIndex const>, std::shared_ptr<DB::IMergeTreeDataPart const>, unsigned long, DB::MarkRanges const&, DB::MarkCache*, DB::UncompressedCache*, DB::MergeTreeReaderSettings) @ 0x0000000014684e6e in /usr/bin/clickhouse
7. DB::MergeTreeDataSelectExecutor::filterMarksUsingIndex(std::shared_ptr<DB::IMergeTreeIndex const>, std::shared_ptr<DB::IMergeTreeIndexCondition>, std::shared_ptr<DB::IMergeTreeDataPart const>, DB::MarkRanges const&, DB::Settings const&, DB::MergeTreeReaderSettings const&, unsigned long&, DB::MarkCache*, DB::UncompressedCache*, Poco::Logger*) @ 0x000000001466a088 in /usr/bin/clickhouse
8. ? @ 0x0000000014666f9a in /usr/bin/clickhouse
9. ? @ 0x0000000014671bfb in /usr/bin/clickhouse
10. ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::worker(std::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>) @ 0x000000000e19be23 in /usr/bin/clickhouse
11. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x000000000e19e615 in /usr/bin/clickhouse
12. ThreadPoolImpl<std::thread>::worker(std::__list_iterator<std::thread, void*>) @ 0x000000000e197cb4 in /usr/bin/clickhouse
13. ? @ 0x000000000e19d4a1 in /usr/bin/clickhouse
14. ? @ 0x00007fe65ed00609 in ?
15. clone @ 0x00007fe65ec25133 in ?
```

<details>
<summary>addr2line</summary>

```
# llvm-addr2line-16 -pafiCse ./clickhouse 0x000000000e0c3d95  0x00000000146daa67  0x00000000146d9f22  0x00000000146d80ea  0x00000000146d7dd8 0x00000000146ff601  0x0000000014684e6e 0x000000001466a088  0x0000000014666f9a
0xe0c3d95: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) at Exception.cpp:92
0x146daa67: std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::__is_long[abi:v15000]() const at string:1499
 (inlined by) std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::~basic_string() at string:2333
 (inlined by) DB::Exception::MessageMasked::~MessageMasked() at Exception.h:42
 (inlined by) DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&&, int, bool) at Exception.h:54
 (inlined by) DB::Exception::Exception<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&, unsigned long&, unsigned long&, unsigned long&, unsigned long>(int, FormatStringHelperImpl<std::__1::type_identity<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&>::type, std::__1::type_identity<unsigned long&>::type, std::__1::type_identity<unsigned long&>::type, std::__1::type_identity<unsigned long&>::type, std::__1::type_identity<unsigned long>::type>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&, unsigned long&, unsigned long&, unsigned long&, unsigned long&&) at Exception.h:81
0x146d9f22: DB::MergeTreeMarksLoader::loadMarksImpl() at MergeTreeMarksLoader.cpp:163
0x146d80ea: DB::MergeTreeMarksLoader::loadMarks() at MergeTreeMarksLoader.cpp:200
0x146d7dd8: std::__1::enable_if<is_move_constructible<DB::MarksInCompressedFile*>::value && is_move_assignable<DB::MarksInCompressedFile*>::value, void>::type std::__1::swap[abi:v15000]<DB::MarksInCompressedFile*>(DB::MarksInCompressedFile*&, DB::MarksInCompressedFile*&) at swap.h:37
 (inlined by) std::__1::shared_ptr<DB::MarksInCompressedFile>::swap[abi:v15000](std::__1::shared_ptr<DB::MarksInCompressedFile>&) at shared_ptr.h:761
 (inlined by) std::__1::shared_ptr<DB::MarksInCompressedFile>::operator=[abi:v15000](std::__1::shared_ptr<DB::MarksInCompressedFile>&&) at shared_ptr.h:723
 (inlined by) DB::MergeTreeMarksLoader::getMark(unsigned long, unsigned long) at MergeTreeMarksLoader.cpp:80
0x146ff601: DB::MergeTreeReaderStream::init() at MergeTreeReaderStream.cpp:72
0x14684e6e: DB::MergeTreeReaderStream::adjustRightMark(unsigned long) at MergeTreeReaderStream.cpp:276
 (inlined by) DB::MergeTreeIndexReader::MergeTreeIndexReader(std::__1::shared_ptr<DB::IMergeTreeIndex const>, std::__1::shared_ptr<DB::IMergeTreeDataPart const>, unsigned long, DB::MarkRanges const&, DB::MarkCache*, DB::UncompressedCache*, DB::MergeTreeReaderSettings) at MergeTreeIndexReader.cpp:61
0x1466a088: std::__1::__deque_base<DB::MarkRange, std::__1::allocator<DB::MarkRange>>::__deque_base() at deque:1188
 (inlined by) std::__1::deque<DB::MarkRange, std::__1::allocator<DB::MarkRange>>::deque[abi:v15000]() at deque:1313
 (inlined by) DB::MarkRanges::MarkRanges() at MarkRange.h:34
 (inlined by) DB::MergeTreeDataSelectExecutor::filterMarksUsingIndex(std::__1::shared_ptr<DB::IMergeTreeIndex const>, std::__1::shared_ptr<DB::IMergeTreeIndexCondition>, std::__1::shared_ptr<DB::IMergeTreeDataPart const>, DB::MarkRanges const&, DB::Settings const&, DB::MergeTreeReaderSettings const&, unsigned long&, DB::MarkCache*, DB::UncompressedCache*, Poco::Logger*) at MergeTreeDataSelectExecutor.cpp:1694
0x14666f9a: DB::MergeTreeDataSelectExecutor::filterPartsByPrimaryKeyAndSkipIndexes(std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const>>>&&, std::__1::vector<std::__1::shared_ptr<DB::AlterConversions const>, std::__1::allocator<std::__1::shared_ptr<DB::AlterConversions const>>>&&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const>, DB::SelectQueryInfo const&, std::__1::shared_ptr<DB::Context const> const&, DB::KeyCondition const&, DB::MergeTreeReaderSettings const&, Poco::Logger*, unsigned long, std::__1::vector<DB::ReadFromMergeTree::IndexStat, std::__1::allocator<DB::ReadFromMergeTree::IndexStat>>&, bool)::$_1::operator()(unsigned long) const at MergeTreeDataSelectExecutor.cpp:1071

```

</details>

### LINKED ISSUE #53858: DB::Exception: Too many marks in file skip_idx_sindex_visitorid.cmrk3, marks expected 3 (bytes size 72)
Hi,

I'm try to use the latest version 23.7.4.5 (official build) and skip index, but after I insert data and try to query via skipe index, I got this error, can you please tell me why?

I get started with clickhouse but stuck at first step here

`error: HttpCode:500 ;  ;Code: 33. DB::Exception: Too many marks in file skp_idx_sindex_visitorid.cmrk3, marks expected 3 (bytes size 72). (CANNOT_READ_ALL_DATA) (version 23.7.4.5 (official build))`

Thank you very much!

### PR DESCRIPTION
**Pull request by canhld94**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Fix recalculation of skip indexes in ALTER UPDATE queries when table has adaptive granularity

When table has adaptive granularity, `ALTER UPDATE` may change number of marks in compact parts.

Reproduce: https://fiddle.clickhouse.com/9f2b67f0-07ab-4d90-82fa-5f2c1abf7ee7

```sql
CREATE TABLE kv
(
    `key` UInt64,
    `value` UInt64,
    `s` String,
    INDEX value_idx value TYPE SET(0) GRANULARITY 1
)
ENGINE = ReplacingMergeTree
ORDER BY key
SETTINGS index_granularity = 32, index_granularity_bytes = 1024;

INSERT INTO kv SELECT
    number,
    number + 100,
    toString(number)
FROM numbers(2048);

ALTER TABLE kv
    UPDATE s = 'barzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzdgfasdghasldfasdfaghfweaufghdasgjasd;lfkaksvnasdljfhjd;alkfgjsdaghasdhgfaosdfhdasjh' WHERE 1
SETTINGS mutations_sync = 2;

SELECT *
FROM kv
WHERE value = 42;
```

Certainly close #53858, close #52652 

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by amosbird** [comment]
Why do we need to rebuild projections when the granularity of parent part change?

**Comment by canhld94** [comment]
> Why do we need to rebuild projections when the granularity of parent part change?

@amosbird I didn't thought much about projections when making this changes. It seem we don't need to rebuild projections in this case. I will push a fix.

Also cc @CurtizJ how do you think about this changes?

**Comment by canhld94** [comment]
Yes, projections have no issue https://fiddle.clickhouse.com/379c0cc1-92de-47f2-a038-0aa90de07521

File: src/Interpreters/MutationsInterpreter.cpp [hunk 1]
```diff
@@ -687,6 +695,9 @@ void MutationsInterpreter::prepare(bool dry_run)
                     }
                 }
             }
+
+            if (source.isCompactPart() && source.getMergeTreeData() && source.getMergeTreeData()->getSettings()->index_granularity_bytes > 0)
+                need_rebuild_indexes = true;
```
Comment (CurtizJ) [inline src/Interpreters/MutationsInterpreter.cpp:700]: Please, add a comment.

**Review by CurtizJ — COMMENTED** [review]
It looks correct. But to be honest I already barely understand the code near recalculations of skip indexes and other dependencies. Probably we have to refactor it a bit in the future.

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit a09c53233b5f9bb97104681a62b2d7021ea6a63a with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55202/a09c53233b5f9bb97104681a62b2d7021ea6a63a/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6387503045">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/55202/a09c53233b5f9bb97104681a62b2d7021ea6a63a/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by amosbird** [comment]
> Yes, projections have no issue

The title and change log entry also need update.

**Review by CurtizJ — APPROVED** [review]

**Pull request merged**